### ISSUE TITLE
Our inheritance is screwed up

### ISSUE THREAD

**Issue by Leont** [Perl-Toolchain-Gang/ExtUtils-MakeMaker#91]
We have a historical split of most of our functionality between MM_Any and MM_Unix, the latter being a subclass of the former. However, all currently used MM_\* classes are subclasses of MM_Unix, which suggests that this distinction is obsolete (MM_MacOS was an example of a direct MM_Any subclass).

To make things more complicated, various subclasses for non-unix operating systems(Win32, VMS, DOS, BeOS) are `@ISA = (MM_Any, MM_Unix)`, which is a rather peculiar order. AFAIK that only matter for 6 methods: all_target, init_platform, init_tools, platform_constants, test_via_harness, test_via_script.

I suspect we could move all of MM_Unix into MM_Any except those methods, and would end up with a simpler system. YMMV.

**Comment by mohawk2** [comment]
Note to whomever does this: File::Find returns paths /-separated, while File::Spec obviously is FS-dependent. This can cause mismatches in make targets.

**Comment by mohawk2** [comment]
Another issue is EU::Liblist[::Kid] are a bit "challenging", and the code should instead be incorporated into MM_*.

**Comment by Leont** [comment]
> Another issue is EU::Liblist[::Kid] are a bit "challenging", and the code should instead be incorporated into MM_*.

Yeah, that code is just really confused if you ask me. I can't imagine why it is the way it is.

**Comment by Leont** [comment]
Come to think about this, this may break a few Makefile.PLs that peek/poke into MM_Unix. A cpan grep suggests this is mainly people calling ExtUtils::MM_Unix::parse_version directly, so if we explicitly import that from MM_Any to MM_Unix

This may affect [HPUX-SDUX](https://metacpan.org/source/ASARIH/HPUX-SDUX-0.03/SDUX.pm#L228) (failure to use `SUPER::`), [DBD::Solid](https://metacpan.org/source/JSLAGEL/DBD-Solid-0.20a/Makefile.PL#L128) (another failure to use `SUPER::`) and [Test::File](https://metacpan.org/source/BDFOY/Test-File-1.44/Makefile.PL#L1) (failure to use `@Overridable`).

**Comment by rurban** [comment]
FYI: The current inheritance is broken with c3 for MM_Win32 and MM_VMS.

**Comment by rurban** [comment]
FYI: I started working on that in my c3 branch. 
use mro 'dfs' as fallback is not possible because mro is a XS and EUMM needs to be run with miniperl in CORE.

I can also can get rid of all the insane run-time require ExtUtils::MM_Any, and do proper compile-time inheritance.

**Comment by haarg** [comment]
Here's a few examples of things people have done with the MM_ modules that could get broken by moving methods naively.
https://metacpan.org/source/ASARIH/HPUX-SDUX-0.03/SDUX.pm#L228,236
https://metacpan.org/source/RYBSKEJ/forks-0.36/Makefile.PL#L128-130
https://metacpan.org/source/JONALLEN/a2pdf-1.13/Makefile.PL#L10
https://metacpan.org/source/JSLAGEL/DBD-Solid-0.20a/Makefile.PL#L128

**Comment by rurban** [comment]
Thanks, I'll test those also.
I'm basically done, but broke parse_version

**Comment by Leont** [comment]
All of those can be easily fixed by extending using `MY::` instead of monkey-patching, and using `MM->parse_version()` instead of accessing MM_Unix directly. We should throw out some pull requests.

**Comment by Leont** [comment]
Yeah, a round of PRs would be necessary. `MM_Unix` actually warns that methods could move to `MM_Any`, at least that part is already covered.
